TFEngine.py

```diff
@@ -1344,48 +1345,84 @@ def _do_save(self):
     See also :func:`self.search` for performing beam search.
 
     :param str|None output_file: if given, will save the results to this file (total err/score for each dataset)
     :param str|None output_per_seq_file: if given, will save the err/score for each sequence
+    :param [str]|None loss_name=None: Needed if output_file is given.
+    :param str|None output_per_seq_format="seq_len;score;pos_score":
+                                            which properies of `loss_name` should be written to `output_per_seq_file`.
+                                            allowed_outputs = {"seq_tags", "seq_len", "score", "error", "pos_score", "pos_error"}
+                                            requested outputs must be separated by a colon ";", e.g. "seq_len;score;error"
+
     :return: nothing
     """
+
     extra_fetches = None
-    results_per_seq = {}  # seq_tag -> dict[str,float]
+    requested_outputs=[]
+
     if output_per_seq_file:
-      extra_fetches = {}
+      allowed_outputs = {"seq_tags", "seq_len", "score", "error", "pos_score", "pos_error"}
+
+      requested_outputs = output_per_seq_format.split(';')
+      assert requested_outputs
+      assert set(requested_outputs) - allowed_outputs == set(), \
+        "Only %s are allowed in function eval_model as output_per_seq_format, but got: %s " % (allowed_outputs, output_per_seq_format)
+
+      # always fetch seq_tag to map loss values to the corresponding line
+      extra_fetches = {"seq_tags": self.network.get_seq_tags()}
+
       from TFUtil import identity
       losses_dict, _, _ = self.network.get_losses_initialized(reduce_func=identity, with_total=False)
-      for loss_name, loss_holder in losses_dict.items():
-        loss_holder.loss.layer = None  # enforce reinit
-        loss_holder.loss.use_flatten_frames = False  # we need that such that we get (B*T,...) unreduced values
-        loss_value = loss_holder.get_normalized_loss_value_per_seq()
-        error_value = loss_holder.get_normalized_error_value_per_seq()
-        if loss_value is not None:
-          extra_fetches["loss:score:%s" % loss_name] = loss_value
-        if error_value is not None:
-          extra_fetches["loss:error:%s" % loss_name] = error_value
-        if loss_holder.loss.output.have_time_axis():
-          seq_lens = loss_holder.loss.output.get_sequence_lengths()
-          extra_fetches["loss:seq_lens:%s" % loss_name] = seq_lens
-        extra_fetches["seq_tags"] = self.network.get_seq_tags()
-      assert extra_fetches
-
+      assert loss_name in losses_dict, \
+        "Unknown loss defined. Got \"%s\". Possible losses are %s" % (loss_name, losses_dict.keys())
+
+      #for loss_name, loss_holder in losses_dict.items():
+      loss_holder = losses_dict[loss_name]
+      loss_holder.loss.layer = None                  # enforce reinit, otherwise the new value of 'loss_holder.loss.use_flatten_frames' will be ignored
```

Comment:
The number of spaces between some Python statement and the comment should be exactly 2.
Also, this line looks too long. In such cases, I would put the comment just in the line before the Python statement.